### PR TITLE
ENH: Use strongly typed enums

### PR DESCRIPTION
**Pull request by jhlegarreta**
Use strongly type enums.

### CODE REVIEW

**Comment by jhlegarreta** [comment]
@dzenanz @thewtex This fails to link on my Windows machine (VS 2015). Not sure what is missing in the implementation. It points to the `SetSimilarityMetric` method (macro'ed). Not sure if the underlying cause is something having to do with the `extern` include and print method definition. Any idea?

**Comment by ntustison** [comment]
Awesome @jhlegarreta .  Thanks.

**Comment by jhlegarreta** [comment]
> @dzenanz @thewtex This fails to link on my Windows machine (VS 2015). Not sure what is missing in the implementation. It points to the SetSimilarityMetric method (macro'ed). Not sure if the underlying cause is something having to do with the extern include and print method definition. Any idea?

@dzenanz @thewtex Not sure why some CI builds succeed; the error I get locally seems general enough so as to make any linker fail; my ITK build is also up-to-date and builds fine.

BTW, @ntustison having a look at other things in ITK I've found the `itk::PatchBasedDenoisingImageFilter` and its base class. Not sure if your `itk::NonLocalPatchBasedImageFilter` should rather inherit from one of those and maybe benefit both parts from a better code re-use.

**Comment by ntustison** [comment]
@jhlegarreta Yeah, you might find a way to inherit from that class although I could not.  I am aware of that algorithm and, in fact, I know Suyash (who designed that algorithm) and talked with Kristin(?) a bit as she was implementing it awhile ago.  Unfortunately, it's relatively slow and I didn't see much overlap in the design.  I should point out that I have other classes in ANTs that derive from this class in addition to denoising.  There's the [super-resolution](https://github.com/ANTsX/ANTs/blob/master/Utilities/itkNonLocalSuperresolutionImageFilter.h) class which we don't use much given that our deep learning tools perform much better but the derived [joint label fusion](https://github.com/ANTsX/ANTs/blob/master/ImageSegmentation/itkWeightedVotingFusionImageFilter.h) class is one of the more widely used utilities.  I didn't include it in this module for various reasons.

**Comment by jhlegarreta** [comment]
> @jhlegarreta Yeah, you might find a way to inherit from that class although I could not. I am aware of that algorithm and, in fact, I know Suyash (who designed that algorithm) and talked with Kristin(?) a bit as she was implementing it awhile ago. Unfortunately, it's relatively slow and I didn't see much overlap in the design. I should point out that I have other classes in ANTs that derive from this class in addition to denoising. There's the super-resolution class which we don't use much given that our deep learning tools perform much better but the derived joint label fusion class is one of the more widely used utilities. I didn't include it in this module for various reasons.

OK, have not looked at the implementations in detail; it detaches from your words that it is not straightforward to merge both, so it may left as a future task or have a reminder as an issue in GitHub.

**Comment by jhlegarreta** [comment]
Once this gets settled this one will follow in a separate PR:
https://github.com/jhlegarreta/ITKAdaptiveDenoising/commit/6fde870ad30612397e01cd9c3d678541d82fa195

**Comment by ntustison** [comment]
That's great, @jhlegarreta .  Thanks for doing that.

**Comment by jhlegarreta** [comment]
I have not tried locally to build the wrappings, but the C++ build fails locally for me:

```
Error LNK2019 unresolved external symbol "class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl itk::operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,enum itk::NonLocalPatchBasedImageFilterEnums::SimilarityMetricType)" (??6itk@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@AEAV12@W4SimilarityMetricType@NonLocalPatchBasedImageFilterEnums@0@@Z) referenced in function "int __cdecl itkAdaptiveNonLocalMeansDenoisingImageFilterTest(int,char * * const)" (?itkAdaptiveNonLocalMeansDenoisingImageFilterTest@@YAHHQEAPEAD@Z) AdaptiveDenoisingTestDriver C:\Binaries\itk-adaptive-denoising\test\itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj 1
```

I cannot see where it comes from. Maybe the wrapping error has to do with that, and maybe also because I did not add the class to the command specifying the order of the wrapping.

Not sure how the CI builds succeed.

**Comment by dzenanz** [comment]
Locally, I run into this:
```text
3>itkAdaptiveNonLocalMeansDenoisingImageFilterTest.cxx
3>ITKCommon-5.1.lib(itkEventObject.obj) : error LNK2005: "public: __cdecl itk::AnyEvent::AnyEvent(void)" (??0AnyEvent@itk@@QEAA@XZ) already defined in itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj
3>ITKCommon-5.1.lib(itkEventObject.obj) : error LNK2005: "public: __cdecl itk::StartEvent::StartEvent(void)" (??0StartEvent@itk@@QEAA@XZ) already defined in itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj
3>ITKCommon-5.1.lib(itkEventObject.obj) : error LNK2005: "public: __cdecl itk::EndEvent::EndEvent(void)" (??0EndEvent@itk@@QEAA@XZ) already defined in itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj
3>ITKCommon-5.1.lib(itkEventObject.obj) : error LNK2005: "public: __cdecl itk::ProgressEvent::ProgressEvent(void)" (??0ProgressEvent@itk@@QEAA@XZ) already defined in itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj
3>itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj : error LNK2019: unresolved external symbol "class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl itk::operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,enum itk::NonLocalPatchBasedImageFilterEnums::SimilarityMetricType)" (??6itk@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@AEAV12@W4SimilarityMetricType@NonLocalPatchBasedImageFilterEnums@0@@Z) referenced in function "public: virtual void __cdecl itk::NonLocalPatchBasedImageFilter<class itk::Image<float,2>,class itk::Image<float,2> >::SetSimilarityMetric(enum itk::NonLocalPatchBasedImageFilterEnums::SimilarityMetricType)" (?SetSimilarityMetric@?$NonLocalPatchBasedImageFilter@V?$Image@M$01@itk@@V12@@itk@@UEAAXW4SimilarityMetricType@NonLocalPatchBasedImageFilterEnums@2@@Z)
3>C:\Dev\ITK-git-2015\bin\Debug\AdaptiveDenoisingTestDriver.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by dzenanz** [comment]
My ITK might not be up to date, I am testing some unrelated Python patch. I will provide updated remarks later.

**Comment by dzenanz** [comment]
With up-to-date ITK build, my local C++ build errors go away.

**Comment by jhlegarreta** [comment]
> With up-to-date ITK build, my local C++ build errors go away.

Maybe the difference lies in the compiler or some CMake flags? I have `ITK_LEGACY_REMOVE` to `off` for ITK.

**Comment by dzenanz** [comment]
I have `ITK_LEGACY_REMOVE` ON.

**Comment by dzenanz** [comment]
I get it in debug mode too:

itkAdaptiveNonLocalMeansDenoisingImageFilterTest.obj : error LNK2019: unresolved external symbol "class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl itk::operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,enum itk::NonLocalPatchBasedImageFilterEnums::SimilarityMetricType)" (??6itk@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@AEAV12@W4SimilarityMetricType@NonLocalPatchBasedImageFilterEnums@0@@Z) referenced in function "public: virtual void __cdecl itk::NonLocalPatchBasedImageFilter<class itk::Image<float,2>,class itk::Image<float,2> >::SetSimilarityMetric(enum itk::NonLocalPatchBasedImageFilterEnums::SimilarityMetricType)" (?SetSimilarityMetric@?$NonLocalPatchBasedImageFilter@V?$Image@M$01@itk@@V12@@itk@@UEAAXW4SimilarityMetricType@NonLocalPatchBasedImageFilterEnums@2@@Z)

**Comment by jhlegarreta** [comment]
At least the failure is reproducible. I still do not have an idea where it stems from, though.

**Comment by jhlegarreta** [comment]
@thewtex any chance to have a look at this :pray: ?

**Comment by dzenanz** [comment]
I figured out the problem: `AdaptiveDenoising_LIBRARIES` should be set for external builds too. The solution is in [my fork](https://github.com/dzenanz/ITKAdaptiveDenoising/commit/d701a36e40d71c9d1e96692058ca9101734ccf5f).

**Comment by ntustison** [comment]
Awesome. @dzenanz .  So do you need me to do anything?  Or are the next steps up to somebody else?

**Comment by dzenanz** [comment]
If Jon cherry-picks or squashes my commit into this branch, then you just need to merge the PR. If Jon does not react within a few days, I can open a new PR with my branch which includes the commit from this PR.

**Comment by ntustison** [comment]
Sounds good.  Thanks.

**Comment by jhlegarreta** [comment]
@dzenanz :100: 🏅 Kudos. Thanks for having had a look at this ! I will cherry-pick the commit into this branch.

**Comment by jhlegarreta** [comment]
Looks like the Python packaging is failing, e.g. at the [build-windows-python-packages (5)](https://github.com/ntustison/ITKAdaptiveDenoising/pull/2/checks?check_run_id=1303271562) build:
```
FAILED: C:/P/IPP/ITK-win_35-x64/Wrapping/itkNonLocalPatchBasedImageFilterEnums.xml 
cmd.exe /C "cd /D D:\a\im\_skbuild\win-amd64-3.5\cmake-build\Wrapping\Modules\AdaptiveDenoising &&
D:\a\im\_skbuild\win-amd64-3.5\cmake-build\Wrapping\Generators\CastXML\castxml\bin\castxml.exe
-o C:/P/IPP/ITK-win_35-x64/Wrapping/itkNonLocalPatchBasedImageFilterEnums.xml
--castxml-gccxml --castxml-start _wrapping_
--castxml-cc-msvc
( "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe"
/DWIN32 /D_WINDOWS /W3 /GR /EHsc /bigobj /bigobj )
-fexceptions -w -c @C:/P/IPP/ITK-win_35-x64/Wrapping/AdaptiveDenoising.castxml.inc
C:/P/IPP/ITK-win_35-x64/Wrapping/itkNonLocalPatchBasedImageFilterEnums.cxx"
C:/P/IPP/ITK-win_35-x64/Wrapping/itkNonLocalPatchBasedImageFilterEnums.cxx:16:10: fatal error:
'itkNonLocalPatchBasedImageFilterEnums.h' file not found
#include "itkNonLocalPatchBasedImageFilterEnums.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

I am wondering whether I should remove the `wrapping/itkNonLocalPatchBasedImageFilterEnums.wrap` file, and add the line it contains to the `wrapping/itkNonLocalPatchBasedImageFilter.wrap` file just before it current contents.

**Comment by jhlegarreta** [comment]
> I am wondering whether I should remove the wrapping/itkNonLocalPatchBasedImageFilterEnums.wrap file, and add the line it contains to the wrapping/itkNonLocalPatchBasedImageFilter.wrap file just before its current contents.

I push forced that and it still fails with the same error, so it is something else.

**Comment by jhlegarreta** [comment]
So I finally found the way to address the Python packaging issues ✔️ : cross-referencing the Matt's PR in ITK that allowed to address the issue for the sake of completeness: https://github.com/InsightSoftwareConsortium/ITK/pull/1353

So the PR is ready to be merged. Thanks for the patience and help.

File: include/itkNonLocalPatchBasedImageFilter.h [hunk 1]
```diff
@@ -78,13 +104,13 @@ class ITK_TEMPLATE_EXPORT NonLocalPatchBasedImageFilter :
...
   using NeighborhoodOffsetListType = std::vector<NeighborhoodOffsetType>;
-  /**
-   * Neighborhood patch similarity metric enumerated type
-   */
-  enum SimilarityMetricType {
-    PEARSON_CORRELATION,
-    MEAN_SQUARES
-  };
+
+  using SimilarityMetricTypeEnum = NonLocalPatchBasedImageFilterEnums::SimilarityMetric;
```
Comment (dzenanz) [inline include/itkNonLocalPatchBasedImageFilter.h:108]: should this be `SimilarityMetricEnum` (without `Type` in the name)?

**Review by dzenanz — COMMENTED** [review]
I think `type` is generally replaced by `enum` in new style. See in-line comment.

For cleaner history, you should squash my commit into yours. I made a commit as it was easier to convey my message that way (less writing, no chance to misunderstand).

File: include/itkNonLocalPatchBasedImageFilter.h (hunk 1 above)
Comment (jhlegarreta) [reply include/itkNonLocalPatchBasedImageFilter.h]: :eyes: Agreed ! Thanks for pointing this out.  Done in bf8a591.

**Comment by jhlegarreta** [comment]
> For cleaner history, you should squash my commit into yours. I made a commit as it was easier to convey my message that way (less writing, no chance to misunderstand).

Squashed in bf8a591. Thanks @dzenanz :+1:.

**Review by dzenanz — APPROVED** [review]
Outstanding work!

**Comment by ntustison** [comment]
Thanks.

**Pull request merged**